File: ruby/word-count/word_count.rb
class Phrase

  VERSION = 1

  def initialize(sentence)
    @words = sentence.split(/[ ,\n]/)
  end

  def word_count
    counts = Hash.new(0)
    @words.each do |word|
      normalized_word = normalize(word)
      counts[normalized_word] += 1 if normalized_word.length > 0
    end
    counts
  end

  private

  def normalize(word)
    word.downcase!
    word.sub! /^'/, ''
    word.sub! /[^a-z0-9]+$/, ''
    word
  end

end

File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t) :: map()
  def count(sentence) do
    no_split_for_letters_numbers_and_dashes = ~r/([^\p{L}\p{N}\p{Pd}])+/u

    sentence
    |> String.downcase
    |> String.split(no_split_for_letters_numbers_and_dashes, trim: true)
    |> Enum.reduce(Map.new, &accumulate_count/2)
  end

  defp accumulate_count(word, dict) do
    Map.update(dict, word, 1, &Enum.sum([1, &1]))
  end
end
